### PR TITLE
[task][s2i] - v0.3: option to provide container registry credentials in a workspace

### PR DESCRIPTION
**Pull request by zroubalik**
Signed-off-by: Zbynek Roubalik <zroubalik@gmail.com>

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Follows the [authoring recommendations](https://github.com/tektoncd/catalog/blob/main/recommendations.md)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality of task changed or new task added)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Complies with [Catalog Organization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [x] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [x] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [x] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [x] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [x] mandatory `spec.description` follows the convention

          ```

          spec:
            description: >-
              one line summary of the resource

              Paragraph(s) to describe the resource.
          ```

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

---

[TEP]: https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md
[example]: https://github.com/tektoncd/catalog/tree/master/task/git-clone/0.1
[validation]:  https://github.com/tektoncd/catalog/issues/413

### CODE REVIEW

**Review by vdemeester — APPROVED** [review]

File: task/s2i/0.3/s2i.yaml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+          name: varlibcontainers
+        - mountPath: /gen-source
+          name: gen-source
+    - name: push
+      image: quay.io/buildah/stable:v1.17.0
+      script: |
+        [[ "$(workspaces.sslcertdir.bound)" == "true" ]] && CERT_DIR_FLAG="--cert-dir $(workspaces.sslcertdir.path)"
+        buildah ${CERT_DIR_FLAG} push --tls-verify=$(params.TLSVERIFY) --digestfile $(workspaces.source.path)/image-digest \
+        $(params.IMAGE_NAME) docker://$(params.IMAGE_NAME)
+      env:
```
Comment (piyush-garg) [inline task/s2i/0.3/s2i.yaml:84]: Should we set the env based on the workspace bound?

File: task/s2i/0.3/README.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+- **PATH_CONTEXT**: Source path from where s2i command need to be run
+  (_default: ._).
+- **TLSVERIFY**: Verify the TLS on the registry endpoint (for push/pull to a
+  non-TLS registry) (_default:_ `true`)
+
+## Workspaces
+
+- **source**: A [Workspace](https://github.com/tektoncd/pipeline/blob/main/docs/workspaces.md) containing the source to build.
+- **sslcertdir**: An [_optional_ Workspace](https://github.com/tektoncd/pipeline/blob/v0.17.0/docs/workspaces.md#optional-workspaces) containing your custom SSL certificates to connect to the registry. Buildah will look for files ending with \*.crt, \*.cert, \*.key into this workspace. See [this sample](./samples/openshift-internal-registry.yaml) for a complete example on how to use it with OpenShift internal registry.
+ - **`dockerconfig`**: An [optional workspace](https://github.com/tektoncd/pipeline/blob/main/docs/workspaces.md#using-workspaces-in-tasks)
```
Comment (piyush-garg) [inline task/s2i/0.3/README.md:30]: indentation in workspace list

File: task/s2i/0.3/s2i.yaml (hunk 1 above)
Comment (zroubalik) [reply task/s2i/0.3/s2i.yaml:84]: @piyush-garg I am happy to update this PR, though I am not sure what are you asking for :) Could you please give me some pointers how to do that?

File: task/s2i/0.3/README.md (hunk 2 above)
Comment (zroubalik) [reply task/s2i/0.3/README.md:30]: Good catch, you are right!

File: task/s2i/0.3/s2i.yaml (hunk 1 above)
Comment (vdemeester) [reply task/s2i/0.3/s2i.yaml:84]: I guess it's the same as https://github.com/tektoncd/operator/pull/922#issuecomment-1195138573

Comment (zroubalik) [reply task/s2i/0.3/s2i.yaml]: Fixed in https://github.com/tektoncd/catalog/pull/1025/commits/fb0deae24d014f22c3427965fdc9e9fd6a283c46 PTAL

**Comment by zroubalik** [comment]
Buildah covered in a separate PR: https://github.com/tektoncd/catalog/pull/1027

File: task/s2i/0.3/s2i.yaml [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        [[ "$(workspaces.sslcertdir.bound)" == "true" ]] && CERT_DIR_FLAG="--cert-dir $(workspaces.sslcertdir.path)"
+        [[ "$(workspaces.dockerconfig.bound)" == "true" ]] && export DOCKER_CONFIG="$(workspaces.dockerconfig.path)"
+        buildah ${CERT_DIR_FLAG} push --tls-verify=$(params.TLSVERIFY) --digestfile $(workspaces.source.path)/image-digest \
+        $(params.IMAGE) docker://$(params.IMAGE)
+      securityContext:
+        privileged: true
+      volumeMounts:
+        - mountPath: /var/lib/containers
+          name: varlibcontainers
+    - name: digest-to-results
```
Comment (vinamra28) [inline task/s2i/0.3/s2i.yaml:90]: as part of optimization, can we merge all these 3 steps ?

Comment (vinamra28) [reply task/s2i/0.3/s2i.yaml:90]: for reference you can see https://github.com/tektoncd/catalog/blob/5e1f510bb47020585cb8277adc4ad903b2aaa8ba/task/buildah/0.4/buildah.yaml#L65-L80

Comment (zroubalik) [reply task/s2i/0.3/s2i.yaml:90]: okay

Comment (zroubalik) [reply task/s2i/0.3/s2i.yaml]: @vinamra28 done, PTAL

**Comment by zroubalik** [comment]
@piyush-garg @vdemeester any chance getting this merged? 🙇‍♂️

**Comment by vinamra28** [comment]
PR looks good. @zroubalik can you please squash the commits?

**Comment by zroubalik** [comment]
> PR looks good. @zroubalik can you please squash the commits?

@vinamra28 done

**Comment by vinamra28** [comment]
/lgtm
Thanks for your contribution 😄

**Comment by zroubalik** [comment]
So, can we merge this one? :)

**Pull request merged**